src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +149,170 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+    assert(_collected_nodes.size() == 0, "should not call this method twice in a row");
+    assert(!_is_target_node(start_node), "no trivial paths where start node is also a target");
+
+    push_unvisited_node(start_node);
+    while (_stack.is_nonempty()) {
+      Node* current = _stack.node();
+      _visited.set(current->_idx);
+      if (_is_target_node(current)) {
+        // Target node? Do not visit its inputs and begin backtracking.
+        _collected_nodes.push(current);
+        pop_target_node_and_collect_predecessor();
+      } else if (!push_next_unvisited_input()) {
+        // All inputs visited. Continue backtracking.
+        pop_node_and_maybe_collect_predecessor();
+      }
+    }
+    return _collected_nodes;
+  }
+
+ private:
+  // The predecessor (just below the target node (currently on top) on the stack) is also on the path from
+  // start->target. Collect it and pop the target node from the top of the stack.
+  void pop_target_node_and_collect_predecessor() {
+    _stack.pop();
+    assert(_stack.is_nonempty(), "target nodes should not be start nodes");
+    _collected_nodes.push(_stack.node());
+  }
+
+  // Push the next unvisited input node of the current node on top of the stack by using its stored associated input index:
+  //
+  //                        Stack:
+  //      I1  I2  I3        [current, 2] // Index 2 means that I1 (first data node at index 1) was visited before.
+  //       \  |  /                       // The next unvisited input is I2. Visit I2 by pushing a new entry [I2, 1]
+  //   Y   current                       // and update the index [current, 2] -> [current, 3] to visit I3 once 'current'
+  //    \  /                             // is on top of stack again later in DFS walk.
+  //     X                  [X, 3]       // Index 3 points past node input array which means that there are no more inputs
+  //                                     // to visit. Once X is on top of stack again, we are done with 'X' and pop it.
+  //
+  // If an input was already collected before (i.e. part of start->target), then the current node is part of some kind
+  // of diamond with it:
+  //
+  //        C3       X
+  //       /  \     /
+  //     C2   current
+  //      \  /
+  //       C1
+  //
+  // Cx means collected. Since C3 is already collected (and thus already visited), we add 'current' to the collected list
+  // since it must also be on a path from start->target. We continue the DFS with X which could potentially also be on a
+  // start->target path but that is not known yet.
+  //
+  // This method returns true if there is an unvisited input and return false otherwise if all inputs have been visited.
+  bool push_next_unvisited_input() {
+    Node* current = _stack.node();
+    const uint next_unvisited_input_index = _stack.index();
+    for (uint i = next_unvisited_input_index; i < current->req(); i++) {
+      Node* input = current->in(i);
+      if (_node_filter(input)) {
```

Comment:
you could check that if the filter does not pass, then also the target-node criterion fails.